### PR TITLE
[rpc] cpu improvement of protobuf in gcs

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

Make use of rpc arena to avoid protobuf copying and destructing which are really expensive for complicated structures.
This is a split of #17828

## Related issue number

#17805
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
Nice! can you share any measurements/that show the affect of this?

File: src/ray/raylet_client/raylet_client.cc
Comment (rkooo567) [inline src/ray/raylet_client/raylet_client.cc:311]: According to the document, this seems to allocate large memory ahead of time. Isn't it the anti-pattern to create Arena for each RPC call? (i.e. doesn't it need to be a member of RayletClient?)

**Review by rkooo567 — COMMENTED** [review]
I think it'll be also great we introduce this to resource update? 

Also, I have a question about this. According to the doc, it seems like Arena preallocates memory and create requests/replies there so that it avoids memory allocation / destruction cost, but what's going to happen if the preallocated memory is not enough? Is it automatically increased/decreased?

**Review by ericl — COMMENTED** [review]
Unrelated, but I could see this really speeding up core worker and raylet as well.

**Comment by iycheng** [comment]
> Nice! can you share any measurements/that show the affect of this?

I'll try to get some cpu utilization dashboard first to visualize the improvement. I think I made one mistake in the naming. It's actually CPU improvement...

File: src/ray/raylet_client/raylet_client.cc
Comment (iycheng) [reply src/ray/raylet_client/raylet_client.cc:311]: It's actually is a good pattern. Arena will not release allocated block and also it won't reuse the freed memory. 
Also to answer your other question mentioned, it'll auto increase. Internally, it's chained blocks. There is a parameter that we can tune for the size of each block.

When allocate, it'll allocate blocks and in destruction time, it'll free blocks. So protobuf won't trigger any destruction (think about complicated map objects destruction which is expensive).

In the end, all sending requests should all be built on top of the arena I think. For reply, it's not necessary and we should make replies non-const so that we can release them manually and attach them to the main thread structure.

Some references here:
>Granularity
We have found in most application server use cases that an "arena-per-request" model works well. You may be tempted to divide arena use further, either to reduce heap overhead (by destroying smaller arenas more often) or to reduce perceived thread-contention issues. However, the use of more fine-grained arenas may lead to unintended message copying, as we describe above. We have also spent effort to optimize the Arena implementation for the multithreaded use-case, so a single arena should be appropriate for use throughout a request lifetime even if multiple threads process that request.

File: src/ray/gcs/gcs_server/gcs_node_manager.cc
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:86]: Can you add a comment as to why these unsafe adds are OK? (presumably since both messages were allocated in the same arena?)

File: src/ray/raylet_client/raylet_client.cc
Comment (ericl) [inline src/ray/raylet_client/raylet_client.cc:314]: Comment as to why unsafe is ok?

File: src/ray/rpc/server_call.h
Comment (ericl) [inline src/ray/rpc/server_call.h:230]: Comment about lifetime?

File: src/ray/rpc/server_call.h
Comment (ericl) [inline src/ray/rpc/server_call.h:258]: Comment that it's owned by the arena?

**Review by ericl — APPROVED** [review]

File: src/ray/raylet_client/raylet_client.cc
Comment (rkooo567) [reply src/ray/raylet_client/raylet_client.cc:311]: I see. So in this case, the deallocation will be faster (since it is just freeing blocks), but we still need to allocate memory (a list of blocks) for each request creation time, am I correct?

**Review by rkooo567 — APPROVED** [review]
LGTM. Btw, I am also curious about memory overhead. Would this introduce noticeable additional memory usage? Is it possible to see if there's high memory overhead in our 10K actor tests?

File: src/ray/raylet_client/raylet_client.cc
Comment (wuisawesome) [reply src/ray/raylet_client/raylet_client.cc:311]: I think allocation should be faster too since we can do a single allocation (and there's less fragmentation)? 

We should probably think about the per-request overhead we're introducing, but let's say we have a 100k actor scalability envelope and we allocate 10KiB (this is probably high) per actor, that's only 1GiB of overhead. 

We could probably measure this more carefully, but back of napkin estimate says it's probably a good tradeoff for us?

Comment (rkooo567) [reply src/ray/raylet_client/raylet_client.cc:311]: Yeah agree that degree of tradeoff is worth it. One concern I have is that we might use this more often than actor creation path, and the accumulated overhead could be very large (in that case, we need some tuning like reducing block size), and the memory usage regression is really hard to detect & debug, so it’d be still great to understand the accurate overhead? Wdyt

Comment (rkooo567) [reply src/ray/raylet_client/raylet_client.cc:311]: Maybe if we can calcuate that, there’s no need to do actual benchmark

Comment (iycheng) [reply src/ray/raylet_client/raylet_client.cc:311]: @rkooo567  @wuisawesome  I think it depends. First of all, it's allocating by blocks, and each block is something configurable. The next thing is that sometimes it'll use less memory since for this model, it's able to reference things, not in the request.
You can check the implementation of GetAllNodeInfo.

```
reply->add_node_info_list()->CopyFrom(*entry.second);
```
got changed to
```
    reply->mutable_node_info_list()->UnsafeArenaAddAllocated(entry.second.get());
```

In this way, we avoid copying/constructing/deconstructing, so it benefits both memory and CPU actually.

The default max block size is 8k (https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/arena.h#L131-L169) and the waste/gain depends on the actual request.

I'll measure it before making it ready to be merged.

**Review by wuisawesome — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_node_manager.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_node_manager.cc:86]: Actually, not because of that. It's because entry's life cycle is longer than reply. I'll comment on it there.

**Comment by WangTaoTheTonic** [comment]
Nice improvement!

**Comment by WangTaoTheTonic** [comment]
We also found lots of `GetAllActors` requests in actor launching, can we use this change to that request?

**Comment by iycheng** [comment]
Measuring result here:
Memory: 4.7g (this one)  vs  4.0g (master) (1k nodes), so 700MB more. But it's stable. I guess that's because the block size is bigger for some requests, especially for the ones like an empty request.

The CPU percentage of HandleGetAllNodeInfo drop from 26% -> 11%. 

But I think we probably want to hold this PR for a while until I have a small fix for the scheduling issues. :(
Since it improves the gcs's performance, which makes the conflict worse due to this (https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_actor_scheduler.cc#L435-L440).

**Comment by iycheng** [comment]
@WangTaoTheTonic  do you mean GetAllActorInfo ? Can you show me some code point?

**Comment by WangTaoTheTonic** [comment]
@iycheng Yeah. Like here https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L200

**Comment by iycheng** [comment]
> @iycheng Yeah. Like here https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L200

@WangTaoTheTonic That's right! Since I'm not merging it until scheduling issues are fixed (maybe later this week), I probably will just update it in this PR.

**Comment by ericl** [comment]
Is this still planned to merge?

**Comment by WangTaoTheTonic** [comment]
We run a case and found it will reduce cpu usage obviously too, so it has already been merged into our internal repo. Thanks!
P.S. The simple case is starting a cluster with n(which is large) nodes, then kill GCS and let it start from failure, observe the cpu usage while restarting. (because every node will send a GetAllNodes request after reconnecting to GCS)

**Comment by iycheng** [comment]
> Is this still planned to merge?

@ericl I observed that simply merge this one will cause some actor scheduling issues. So the plan here is:

1. merge actor async submission
2. test this one + simple scheduling change in 1k nodes
3. merge these + scheduling change

In my mind, the timeline should be within one week.

File: src/ray/raylet_client/raylet_client.cc
Comment (mwtian) [inline src/ray/raylet_client/raylet_client.cc:303]: Using Arena here looks good. And if the goal is to save CPU from repeated heap allocation of sub-messages, an alternative is to reuse Message objects by just clearing them in between each usage.
https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.message_lite#MessageLite.Clear.details
If the messages can have large size variances, they can be "thrown away" after 100 uses. It can be more flexible if we also have a lot of preallocated data.

**Review by mwtian — APPROVED** [review]

File: src/ray/raylet_client/raylet_client.cc
Comment (iycheng) [reply src/ray/raylet_client/raylet_client.cc:303]: @mwtian  Thanks for the pointing out. But here the goal is to avoid copy, right now we don't reuse replies around.

**Comment by iycheng** [comment]
I've tested this PR with https://github.com/ray-project/ray/pull/18546 and they work well with each other.

**Pull request merged**
